### PR TITLE
fix patching WorkloadRebalancer failed due to misuse of shared slices

### PR DESCRIPTION
**Pull request by chaosi-zju**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

fix patching WorkloadRebalancer failed due to misuse of shared slices.

in previous code:

```go
// get previous status and update basing on it
newStatus = rebalancer.Status
if len(newStatus.ObservedWorkloads) == 0 {
	newStatus = c.buildWorkloadRebalancerStatus(rebalancer)
}
```

actually `newStatus` and `rebalancer.Status` shared the slice of `rebalancer.Status.ObservedWorkloads`.

if this slice didn't change from nil to non-nil or occurs expansion, modification to `newStatus` will also be applied to the `rebalancer.Status`, which will lead to `patch status` failed. 

**Which issue(s) this PR fixes**:

Fixes part of #4840

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by chaosi-zju** [comment]
@XiShanYongYe-Chang @RainbowMango 

Hello, sorry for this bug, please help review once more.

In previous code:

```go
// get previous status and update basing on it
newStatus = rebalancer.Status
if len(newStatus.ObservedWorkloads) == 0 {
	newStatus = c.buildWorkloadRebalancerStatus(rebalancer)
}
```

Actually `newStatus` and `rebalancer.Status` shared the slice of `rebalancer.Status.ObservedWorkloads`.

If this slice didn't change from nil to non-nil or occurs expansion, modification to `newStatus` will also be applied to the `rebalancer.Status`, which will lead to `patch status` failed.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/lgtm

**Comment by chaosi-zju** [comment]
Is there any further comments? If any other suggestions, please let me know.

**Comment by RainbowMango** [comment]
@XiShanYongYe-Chang 
If you want to get approval from someone else, you can cc or assign.

/assign

**Comment by XiShanYongYe-Chang** [comment]
> If you want to get approval from someone else, you can cc or assign.

Got it.

**Comment by RainbowMango** [comment]
> modification to newStatus will also be applied to the rebalancer.Status, which will lead to patch status failed.

Won't lead to patch failed, but non-ops, right? That means the patch will be ignored.

**Comment by chaosi-zju** [comment]
> Won't lead to patch failed, but non-ops, right? That means the patch will be ignored.

yes, the patch will be ignored.

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (RainbowMango) [inline pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:81]: Not a good pattern to mutate the input parameters, as the method behavior is hard to predict.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Comment by RainbowMango** [comment]
What's the next PR blocked by this？

**Comment by chaosi-zju** [comment]
> What's the next PR blocked by this？

#4860

**Comment by chaosi-zju** [comment]
> What's the next PR blocked by this？

or #4876 if you prefer

**Comment by RainbowMango** [comment]
I can review them but please also ask for review from controller owners.

**Pull request merged**
